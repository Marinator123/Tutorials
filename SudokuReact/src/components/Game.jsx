import React from 'react';
import Board from './Board.jsx';

/**
 * 
 */
class Game extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            squares: instantiateSquares(81)
        }
    }

    setValues(id, value) {
        const squares = this.state.squares.slice();
        // todo check ob squares legit sind
        squares[id].value = value;
        this.setState({ squares: squares })
    }

    render() {
        return (
            <Board
                squares={this.state.squares}
                setValues={(i, j) => this.setValues(i, j)}
            />
        )
    }
}

export default Game;

function instantiateSquares(size) {
    const squares = Array(size).fill({value: '', readOnly: false});
    const percEmptyFields = 0.7;
    for (let i = 0; i < size; i++) {
        if (Math.random() > percEmptyFields) {
            const rowId = Math.floor(i / 9);
            const columnId = i % 9;
            /** Generiert die Startnummer des Felds z.B. 43 -> 33 */
            const startFieldId = i - i % 3 - rowId % 3 * 9;
            const number = findUniqueNumber(rowId, columnId, startFieldId, squares);
            squares[i] = {value: number, readOnly: true}
        }
    }
    return squares;
}

function findUniqueNumber(rowId, columnId, startFieldId, squares) {
    const occupiedNumbers = getAllValuesForRow(rowId, squares).concat(
        getAllValuesForColumn(columnId, squares)).concat(getAllValuesForField(startFieldId, squares))
    while (true) {
        const randomNumber = Math.floor((Math.random() * 9) + 1)
        if (occupiedNumbers.indexOf(randomNumber) === -1) {
            return randomNumber;
        }
    }
}

function getAllValuesForRow(rowId, squares) {
    const rowSize = squares.length / 9;
    const rowSquares = squares.slice(rowId * rowSize, rowId * rowSize + rowSize);
    return rowSquares.map(a => a.value).filter(val => val !== '');
}

function getAllValuesForField(startId, squares) {
    //const startId = (fieldId - fieldId % 3) * 9 + fieldId % 3 * 3;
    const fieldSquares = squares.slice(startId, startId + 3).concat(
        squares.slice(startId + 9, startId + 12).concat(squares.slice(startId + 18, startId + 21)));
    return fieldSquares.map(a => a.value).filter(val => val !== '');
}

function getAllValuesForColumn(columnId, squares) {
    const columnSquares = Array(9).fill(null);
    for (let i = 0; i < columnSquares.length; i++) {
        columnSquares[i] = squares[columnId + i*9];
    }
    return columnSquares.map(a => a.value).filter(val => val !== '');
}