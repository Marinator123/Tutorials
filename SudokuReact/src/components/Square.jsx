import React from 'react';

/**
 * 
 */
class Square extends React.Component {
    constructor(props) {
        super(props);
        
        const regularLineStyle = '1px solid grey';
        const thickLineStyle = '2px solid black';
        this.style = {
            color:this.props.isReadOnly ? 'black' : 'grey',
            borderBottom:[2,5,8].indexOf(this.props.rowId) > -1 ? thickLineStyle: regularLineStyle,
            borderTop:this.props.rowId === 0 ? thickLineStyle: regularLineStyle,
            borderLeft:this.props.id % 9 === 0 ? thickLineStyle: regularLineStyle,
            borderRight:this.props.id % 3===2 ? thickLineStyle: regularLineStyle,
        }

        this.onBlur = this.onBlur.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    onBlur(e) {
        if (!this.props.isReadOnly) {
            var value = e.target.value;
            try {
                validateInput(value);
                if (isNaN(parseInt(value))) value = '';
                else value = parseInt(value);
                this.props.setValues(this.props.id, value);
            } catch (e) {
                this.props.setValues(this.props.id, '');
                alert(e.message);
            }
        }
    }
    
    onChange(e){
            this.props.setValues(this.props.id, e.target.value);
    }

    render () {
        return (
            <input type="text" className="square" value={this.props.value}
                onBlur={this.onBlur} onChange={this.onChange}
                readOnly={this.props.isReadOnly} style={this.style}>
            </input>
        );
    }
}

export default Square;

function validateInput(value) {
    if (value.length > 0) {
        if (isNaN(value)) {
            throw new Error('Input is not a Number');
        } else if (value > 9 || value < 1) {
            throw new Error('Input must be between 1-9')
        }
    }
}