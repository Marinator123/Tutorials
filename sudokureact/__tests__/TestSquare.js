import React from 'react';
import renderer from 'react-test-renderer';
import Square from '../src/components/Square';
import Game from '../src/components/Game';

test('Create new Square', () => {
    const component = renderer.create(
        <Square
                key={1}
                id={1}
                rowId={1}
                value={1}
                isReadOnly={false}
                setValues={null}
            />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Modify a square', () => {
    const component = renderer.create(
        <Game percEmptyFields={1}/>,
    );
    let tree = component.toJSON();
    // Wie ginge das sch√∂ner?
    var exampleSquare = tree.children[0].children[0].children[0].children[0];
    expect(exampleSquare.props.value).toBe('')
    var e = {target:{value:5}};
    exampleSquare.props.onBlur(e);
    tree = component.toJSON();
    exampleSquare = tree.children[0].children[0].children[0].children[0];
    expect(exampleSquare.props.value).toBe(5)
});