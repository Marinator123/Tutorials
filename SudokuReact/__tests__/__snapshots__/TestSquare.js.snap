// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create new Square 1`] = `
<input
  className="square"
  onBlur={[Function]}
  onChange={[Function]}
  readOnly={false}
  style={
    Object {
      "borderBottom": "1px solid grey",
      "borderLeft": "1px solid grey",
      "borderRight": "1px solid grey",
      "borderTop": "1px solid grey",
      "color": "grey",
    }
  }
  type="text"
  value={1}
/>
`;
